include/igl/arap.cpp

```diff
@@ -169,9 +173,9 @@ template <
   typename Derivedbc,
   typename DerivedU>
 IGL_INLINE bool igl::arap_solve(
-  const Eigen::PlainObjectBase<Derivedbc> & bc,
+  const Eigen::MatrixBase<Derivedbc> & bc,
   ARAPData & data,
-  Eigen::PlainObjectBase<DerivedU> & U)
+  Eigen::MatrixBase<DerivedU> & U)
```

Comment:
@alecjacobson I know in our discussion you said outputs should be defined as Eigen::PlainObjectBase in order to allow resizing. However, in this case, I think MatrixBase makes the most sense since the algorithm is not supposed to resize U, and it is convenient again to have the ability to pass an Eigen::Map here.

Reply:
This is simply the convention of libigl. The caller does not need to resize outputs in advance. The cost of this is that we can't use maps for outputs and must copy in situations where the output needs to be written in a caller defined address.

Reply:
```cpp
if(U.size() == 0)
{
// terrible initial guess.. should at least copy input mesh
#ifndef NDEBUG
    cerr<<"arap_solve: Using terrible initial guess for U. Try U = V."<<endl;
#endif
    U = MatrixXd::Zero(data.n,data.dim);
}else
{
    assert(U.cols() == data.dim && "U.cols() match data.dim");
}
```

This snippet of code is why I think PlainObjectBase is doing more harm than good here. It forces the user to provide an object that can be resized and the only case where it needs to be resized is something that is non-sensical algorithmically (or at the very least a terrible initial guess as pointed out by the debug message). So why not instead make the case of U.size() == 0 an error instead? And then there is no requirement for U to be resizable anymore and the user is free to pass in a map to an existing buffer in memory.
I understand if you feel strongly about this in order to make the interface consistent with the rest of libigl but I just thought I'd point this out for reference :)